### ISSUE TITLE
Extracting column vectors of a given table ?

### ISSUE THREAD

**Issue by Allaa-boutaleb** [IBM/tabsketchfm#1]
Hello, first of all, thank you for this innovative and easy-to-follow approach! 

I was playing around with the code for a bit and wanted to experiment with different table union search algorithms in the "online" phase. I was wondering if you plan to release a script that allows us to extract the column vectors of different tables ? Essentially, given a directory of  .csv files, we get a .pkl file that associates each table with its column embeddings, or something like that (kinda similar to how Starmie does it). 

So far, I've managed to train a custom TabSketchFM model using a slightly modified pretraining scripts, but I'm still trying to figure out how I can leverage this model for inference. 

It would be incredibly helpful in what I'm currently investigating. Thanks again!

**Comment by harshakokel** [comment]
Hello @Allaa-boutaleb,

Glad you were able to use the training code. We will make the column-embedding extraction script available by the end of this week. 

Meanwhile, about how to leverage the model for union search. We recommend finetuning the pre-trained model for the task of identifying unionability (in our study we used [LakeBench Union Datasets](https://zenodo.org/records/10042019)) and using embeddings from the finetuned model.

**Comment by Allaa-boutaleb** [comment]
Thanks for getting back to me @harshakokel !

From reading the paper, I was under the impression that finetuning a TabSketchFM model on the LakeBench Union Dataset allows for it to be used as a re-ranker rather than the baseline model for Table Union Search, correct ? 

Essentially, from my understanding, TabSketchFM can be used in two different ways : 

- Either as a table encoder, to obtain contextual column embeddings that can then be leveraged for Table Union Search in the online phase.
- Or as a re-ranker, by finetuning a pretrained TabSketchFM model with a classification head, then using the finetuned model (with the classification head) to rerank results obtained by a baseline model (like Starmie for example).

I'm not sure how I can obtain embeddings from the finetuned model with the classification head. Maybe I need to give the paper another read :p 

Looking forward to the column-embedding extraction script, and thanks again !

**Comment by harshakokel** [comment]
Hello @Allaa-boutaleb, while we did explore use of TabSketchFM as a re-ranker and saw promising results. We found using embeddings from the finetuned model to be more useful for searching directly. 

Extraction of embedding from a finetuned cross-encoder is not conventional. However, our experiments showed that they do contain signal. We are delayed in releasing the code for embedding extraction. My best estimate is that we will be able to release it in Jan.

**Comment by Allaa-boutaleb** [comment]
Hello @harshakokel , Thank you! Can't wait to get my hands on it, this is really cool. 

However, **I have one concern regarding the fine-tuning and evaluation process.** 

For example, for the table union search task: 

Since TUS-SANTOS from LakeBench was built by converting unionable pairs from TUS and SANTOS into a classification dataset, I’m concerned that there might be significant overlap—where the same (query, candidate) unionable pairs used to fine-tune the model also appear in the search benchmarks at test time. Wouldn’t this risk contamination, where the model is essentially tested on pairs it has already seen during fine-tuning (albeit in a different, classification-based format)? 

Could you clarify whether any procedure was implemented to ensure enough disjointness between TUS-SANTOS (used for fine-tuning) and the TUS/SANTOS search benchmarks (used for evaluation)? If no explicit procedure was used, was there any ablation study conducted to measure information leakage - for example, the overlap between positive pairs (label=1) in TUS-SANTOS and the set of all possible unionable table pairs derivable from the TUS and SANTOS search benchmarks?

Thanks again and happy holidays!

**Comment by harshakokel** [comment]
Hello @Allaa-boutaleb ,

We present two sets of evaluation results for search in the [paper](https://arxiv.org/pdf/2407.01619):
1. **Search results** where we used the embedding from the model finetuned on the same dataset and same prediction task  (Sect 3.1–3.3)
2. **Transfer results** where we used embedding from a model finetuned on a different dataset and task. (Sect 3.4)

Perhaps, Section 3.4 would answer your concern? 

Happy Holidays to you too.

**Comment by Allaa-boutaleb** [comment]
Hello @harshakokel ,

I hope you're doing well. Any updates about the script that allows for the extraction of column embeddings ?

Thanks in advance!

**Comment by harshakokel** [comment]
Hello @Allaa-boutaleb ,

The embedding extraction code is available now. Please have a look at the [updated README](https://github.com/IBM/tabsketchfm?tab=readme-ov-file#-search-experiments) for steps.
